File: android/src/main/java/com/ziggeo/tasks/Task.java
package com.ziggeo.tasks;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.facebook.react.bridge.Promise;

import java.util.concurrent.atomic.AtomicInteger;

/**
 * Created by Alex Bedulin on 08.02.2018.
 */

public abstract class Task {

    public static final AtomicInteger GLOBAL_ID = new AtomicInteger();

    protected int id;
    protected Promise promise;
    protected Thread thread;

    public Task(@NonNull Promise promise) {
        this.id = GLOBAL_ID.getAndIncrement();
        this.promise = promise;
    }

    public void setRunnable(Runnable runnable) {
        this.thread = new Thread(runnable);
    }

    public int getId() {
        return id;
    }

    public void execute() {
        thread.start();
    }

    public void resolve(@Nullable Object object) {
        if (promise != null) {
            try {
                promise.resolve(object);
            } finally {
                promise = null;
            }
        }
    }

    public void reject(@NonNull String err) {
        reject(err, "");
    }

    public void reject(@NonNull String err, @Nullable String message) {
        if (promise != null) {
            try {
                promise.reject(err, message);
            } finally {
                promise = null;
            }
        }
    }
}

File: android/src/main/java/com/ziggeo/tasks/Task.kt
package com.ziggeo.tasks

import com.facebook.react.bridge.Promise
import java.util.concurrent.atomic.AtomicInteger

/**
 * Created by Alex Bedulin on 08.02.2018.
 */
open class Task(promise: Promise) {
    private var id: Int
    private var promise: Promise?
    private var thread: Thread? = null
    fun setRunnable(runnable: Runnable?) {
        thread = Thread(runnable)
    }

    fun execute() {
        thread?.start()
    }

    fun resolve(obj: Any?) {
        promise?.let {
            try {
                it.resolve(obj)
            } finally {
                promise = null
            }
        }
    }

    fun reject(err: String) {
        reject(err, null)
    }

    fun reject(err: String, message: String?) {
        promise?.let {
            try {
                it.reject(err, message)
            } finally {
                promise = null
            }
        }
    }

    companion object {
        val GLOBAL_ID = AtomicInteger()
    }

    init {
        id = GLOBAL_ID.getAndIncrement()
        this.promise = promise
    }
}